TransGAN: A Transformer based GAN architecture (#1679)
section-engineering-education/engineering-education: articles/transgan-a-transformer-based-gan-architecture/index.md

At the last stage of the discriminator model, you have a [cls] token that classifies whether the output is real or fake. The CLS token is used for classification tasks in NLP.

Backpropagation is then performed throughout the whole architecture to train the model.

### Tricks for TransGAN

**1. Data augmentation**

The type of data augmentation they use is borrowed from this [paper](https://arxiv.org/pdf/2006.10738.pdf). It highlights different augmentation techniques for GANs. Given that transformers don't have locality bias built into their architecture as CNNs do, they tend to need a lot more data. Data augmentation helps get around this problem by producing more data from the same dataset.

**2. Co-training with Self-Supervised Auxiliary Task**

![Co-training with Self-Supervised Auxiliary Task](/engineering-education/transgan-a-transformer-based-gan-architecture/co-training.PNG)

*[Image Source: Arxiv](https://arxiv.org/pdf/2102.07074.pdf)*

This is a super-resolution task. It is a task performed besides the whole GAN training. 

> The GAN training involves the discriminator model receiving images from both the generator and the real dataset. Through training, this gives us the primary GAN loss.

Reviewer: ```suggestion
The GAN training involves the discriminator model receiving images from both the generator and the real dataset. Through training, we get the primary GAN loss.
```

![Gradually increasing receptive fiels](/engineering-education/transgan-a-transformer-based-gan-architecture/masked-pixel.PNG)

*[Image Source: Arxiv](https://arxiv.org/pdf/2102.07074.pdf)*

### Key takeaways

**1. Model Architecture**

The TransGAN architecture is built purely using transformers. This is unlike previous architectures, which used convolutions or convolutions together with transformers.

**2. Training Techniques**

There are three main techniques employed to train TransGAN better. They include:

- Data augmentation
- Localized initialization for self-attention
- Multi-task co-training for generator with self-supervised auxiliary loss

### Wrapping Up

In this article, we have discussed the TransGAN paper. It attempts to re-design GANs architecture by replacing convolutions with pure transformers.

> Over the past years, we've seen a shift in the use of CNNs to use powerful transformers to perform image classification tasks. Transformers are now being attempted on image generation tasks.  Most of these proposed models still have some challenges. One case is in the use of transformers for image generation tasks such as in GANs. But, they still appear to be very promising architectures. 

Reviewer: ```suggestion
Over time, we have seen a shift in the use of CNNs to use powerful transformers for image classification tasks. Transformers are now being tried on image generation tasks. However, most of these proposed models still pose some challenges. An instance is the use of transformers for image generation tasks, such as in GANs. But, they still appear to be very promising architectures.
```
